src/replication/include/replication/state.hpp

```diff
@@ -21,48 +21,70 @@
 #include "replication/epoch.hpp"
 #include "replication/mode.hpp"
 #include "replication/role.hpp"
+#include "replication_server.hpp"
+#include "status.hpp"
 #include "utils/result.hpp"
 
 namespace memgraph::replication {
 
 enum class RolePersisted : uint8_t { UNKNOWN_OR_NO, YES };
 
+enum class RegisterReplicaError : uint8_t { NAME_EXISTS, END_POINT_EXISTS, COULD_NOT_BE_PERSISTED, NOT_MAIN, SUCCESS };
+
+struct RoleMainData {
+  ReplicationEpoch epoch_;
+  std::vector<ReplicationClientConfig> registered_replicas_;
+};
+
+struct RoleReplicaData {
+  ReplicationServerConfig config;
+  std::unique_ptr<ReplicationServer> server;
+};
+
 struct ReplicationState {
-  ReplicationState(std::optional<std::filesystem::path> durability_dir);
+  explicit ReplicationState(std::optional<std::filesystem::path> durability_dir);
 
   ReplicationState(ReplicationState const &) = delete;
   ReplicationState(ReplicationState &&) = delete;
   ReplicationState &operator=(ReplicationState const &) = delete;
   ReplicationState &operator=(ReplicationState &&) = delete;
 
-  void SetRole(ReplicationRole role) { return replication_role_.store(role); }
-  auto GetRole() const -> ReplicationRole { return replication_role_.load(); }
-  bool IsMain() const { return replication_role_ == ReplicationRole::MAIN; }
-  bool IsReplica() const { return replication_role_ == ReplicationRole::REPLICA; }
-
-  auto GetEpoch() const -> const ReplicationEpoch & { return epoch_; }
-  auto GetEpoch() -> ReplicationEpoch & { return epoch_; }
-
   enum class FetchReplicationError : uint8_t {
     NOTHING_FETCHED,
     PARSE_ERROR,
   };
-  using ReplicationDataReplica = ReplicationServerConfig;
-  using ReplicationDataMain = std::vector<ReplicationClientConfig>;
-  using ReplicationData = std::variant<ReplicationDataMain, ReplicationDataReplica>;
-  using FetchReplicationResult = utils::BasicResult<FetchReplicationError, ReplicationData>;
-  auto FetchReplicationData() -> FetchReplicationResult;
+
+  using ReplicationData_t = std::variant<RoleMainData, RoleReplicaData>;
+  using FetchReplicationResult_t = utils::BasicResult<FetchReplicationError, ReplicationData_t>;
+  auto FetchReplicationData() -> FetchReplicationResult_t;
+
+  auto GetRole() const -> ReplicationRole {
+    return std::holds_alternative<RoleReplicaData>(replication_data_) ? ReplicationRole::REPLICA
+                                                                      : ReplicationRole::MAIN;
+  }
```

Comment:
I don't see a case where this would be an issue, but if we add any other `ReplicationRole` e.g. further specifying replica types if that is ever needed, then this logic will not hold up. Should we static_assert the size of the variant?

Reply:
I see what you are saying about future proofing. Going to not do this for now, mainly because I don't want to put the static_assert in a header.